### ISSUE TITLE
Audio support?

### ISSUE THREAD

**Issue by kevzettler** [giongto35/cloud-morph#37]
Maybe I missed this but is there currently support for audio? The diablo demo has no sound. I'm using the Brave browser.

**Comment by giongto35** [comment]
Yes, Audio is not supported right now : D I need to research how to capture audio in headless machines.

**Comment by Rock-N-Troll** [comment]
Possible via FFMPEG with Linux PulseAudio
https://trac.ffmpeg.org/wiki/Capture/PulseAudio

https://superuser.com/questions/1339038/how-can-i-can-create-a-virtual-pulseaudio-soundcard-on-linux-with-no-audio-hardw

That StackOverflow article also shows PulseAudio can stream over RTP which could prove useful:
https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/User/Modules/#index7h2

**Comment by giongto35** [comment]
Hi @Rock-N-Troll ,
Thanks very much for the resource. The hardest challenge is to capture it when there is no audio hardware. Seems like it's not so complicated with Pulse Audio.
Currently, Im prioritizing to run more applications on the platform. It will be very nice if you can contribute that; or an example about how to capture and get the audio stream at one port is also as nice.

**Comment by Rock-N-Troll** [comment]
@giongto35 
https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/User/Network/RTP/

**"Stream it with vlc as mp3 for low bandwidth"** section
```
load-module module-null-sink sink_name=rtp format=s16be channels=2 rate=44100 description="RTP Multicast Sink"
load-module module-rtp-send source=rtp.monitor destination=127.0.0.1 port=46998 loop=1
```

OR

https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/User/Network/

> How can I use PulseAudio to stream music from my main PC to my LAN with multiple PCs with speakers?
> On the sender side create an RTP sink:
> 
> ```
> load-module module-null-sink sink_name=rtp
> load-module module-rtp-send source=rtp.monitor
> set-default-sink rtp
> ```
> This will make rtp the default sink, i.e. all applications will write to this virtual RTP device by default.
> 

As you can see, it is very straightforward to send audio via RTP if you use PulseAudio and point it to the correct application as the "sink" or the whole Linux environment as the "sink". 

**Potentially can look at Pulse Audio WebRTC repository here from PulseAudio team: https://freedesktop.org/software/pulseaudio/webrtc-audio-processing/**

**Comment by giongto35** [comment]
Just research around today. Seems really promising with PulseAudio. Thanks very much. It helps me a lot indeed! Will try to integrate soon.

**Comment by giongto35** [comment]
@Rock-N-Troll https://github.com/giongto35/cloud-morph/pull/44 Added a simple implementation here, it costs my weekend. 
It works but there is an expected issue that video and audio are not in sync. Will think about it later.
Thanks for giving guides and encouraging me to do it.

**Comment by Rock-N-Troll** [comment]
https://trac.ffmpeg.org/wiki/StreamingGuide

maybe try:
https://stackoverflow.com/questions/16658873/how-to-minimize-the-delay-in-a-live-streaming-with-ffmpeg
-fflags nobuffer
-flags low_delay

(and maybe less important but try a smaller probesize)
-probesize 32

https://ffmpeg.org/ffmpeg-formats.html

can also maybe try alsa instead of pulse https://superuser.com/questions/144648/how-do-alsa-and-pulseaudio-relate
-f alsa 

I think using ffplay to test the delay is the best option to see if it's some part of the stream or the actual audio delay of the whole system:
https://stackoverflow.com/questions/64488707/how-to-solve-the-incremental-delay-in-ffplay-streaming

**Comment by Rock-N-Troll** [comment]
Also, forgot to say how cool that is you were able to get it working!

**Comment by sergystepanov** [comment]
GJ but one small remark.
If you want to use RTP from FFmpeg make sure:
- it generates proper timestamps (delays) for varying video frame times (use_wallclock_as_timestamps?),
- it captures and passes ~20ms audio frames into the Opus encoder without any additional bufferization.
That should be enough for some basic AV-sync with RTP/RTCP/NTP.

Sync [RTFM](https://cdn.ttgtmedia.com/searchUnifiedCommunications/downloads/VideoConf_CH07.pdf).
